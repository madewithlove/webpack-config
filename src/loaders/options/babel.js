export default options => {
    return {
        presets: options.react ? ['react-app', 'stage-0'] : ['env', 'stage-0'],
        plugins: ['transform-decorators-legacy'],
    };
};
