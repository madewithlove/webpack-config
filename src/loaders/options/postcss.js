import autoprefixer from 'autoprefixer';

export default options => ({
    plugins() {
        return [
            autoprefixer({
                browsers: [
                    '>1%',
                    'last 4 versions',
                    'Firefox ESR',
                    'not ie < 9',
                ],
            }),
        ];
    },
});
