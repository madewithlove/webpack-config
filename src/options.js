import path from 'path';
import merge from 'merge';
import loaderOptions from './loaders/options';

export default function (options) {
    const env = process.env.BABEL_ENV || process.env.NODE_ENV || process.env.APP_ENV || 'development';
    const development = typeof options.development === 'undefined'
        ? env !== 'production'
        : options.development;

    options = merge.recursive({

        // Environment
        name: 'main',
        development,
        env,

        // HMR
        domain: process.env.APP_URL,
        hot: options.devServer === false ? development : process.argv.indexOf('--inline') !== -1,

        // Filenames and paths
        filenames: development ? '[name]' : '[name].[hash]',
        devServer: 'http://localhost:8080',
        sourcePath: 'resources/assets/js',
        outputPath: 'public/builds/',

        // Frameworks
        react: true,

        // Other options
        linting: false,
        inlineLimit: 50000,

        // Loaders
        loaders: {
            js: 'babel-loader',
            css: 'css-loader!postcss-loader',
            options: {},
        },

    }, options);

    // Add loader options
    options.loaders.options = merge.recursive(
        loaderOptions(options),
        options.loaders.options,
    );

    // Set NODE_ENV for babel-preset-react-app
    process.env.NODE_ENV = env
        .replace('local', 'development')
        .replace('testing', 'test');

    // Uniformize source path and entry point
    options.entry = options.sourcePath;
    options.sourcePath = path.resolve(path.dirname(options.sourcePath));

    return options;
}
