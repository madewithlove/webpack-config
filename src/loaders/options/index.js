import babel from './babel';
import postcss from './postcss';

export default options => ({
    'babel-loader': babel(options),
    'postcss-loader': postcss(options),
});
