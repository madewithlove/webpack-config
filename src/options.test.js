import options from './options';

it('can merge loader options', () => {
    const merged = options({
        loaders: {
            js: 'foo-loader',
            options: {
                'postcss-loader': {
                    foo: 'bar',
                },
            },
        },
    });

    expect(merged.loaders).toMatchSnapshot();
});
