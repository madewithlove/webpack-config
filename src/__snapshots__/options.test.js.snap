// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can merge loader options 1`] = `
Object {
  "css": "css-loader!postcss-loader",
  "js": "foo-loader",
  "options": Object {
    "babel-loader": Object {
      "plugins": Array [
        "transform-decorators-legacy",
      ],
      "presets": Array [
        "react-app",
        "stage-0",
      ],
    },
    "postcss-loader": Object {
      "foo": "bar",
      "plugins": [Function],
    },
  },
}
`;
